src/state_circuit/state.rs

```diff
@@ -0,0 +1,1241 @@
+|    1    |       89       |  32   |   0  |
+*/
+
+/*
+Example bus mapping:
+
+| global_counter | memory_flag | memory_address | memory_value |
+----------------------------------------------------------------
+|       12       |      1      |        0       |      12      |
+|       17       |      1      |        1       |      32      |
+|       24       |      0      |        0       |      12      |
+|       89       |      0      |        1       |      32      |
+*/
+
+// target:
+// 1 - first row of either target (Note: only the first row, not all init rows)
+// 2 - memory
+// 3 - stack
+// 4 - storage
+
+/// In the state proof, memory operations are ordered first by address, and then by global_counter.
+/// Memory is initialised at 0 for each new address.
+/// Memory is a word-addressed byte array, i.e. a mapping from a 253-bit word -> u8.
+/// TODO: Confirm if we are using 253- or 256-bit memory addresses.
+#[derive(Copy, Clone, Debug)]
+struct Address<F: FieldExt>(F);
+
+/// Global counter
+#[derive(Copy, Clone, Debug)]
+struct GlobalCounter(usize);
+
+/// TODO: In the EVM we can only read memory values in 32 bytes, but can write either
+/// single-byte or 32-byte chunks. In zkEVM:
+/// - Are we reading single bytes or 32-byte chunks? TBD
+/// - We are writing a single field element (253 bits)
+#[derive(Copy, Clone, Debug)]
+struct Value<F: FieldExt>(F);
+
+#[derive(Clone, Debug)]
+enum ReadWrite<F: FieldExt> {
+    // flag == 0
+    Read(GlobalCounter, Value<F>),
+    // flag == 1
+    Write(GlobalCounter, Value<F>),
+}
+
+impl<F: FieldExt> ReadWrite<F> {
+    fn global_counter(&self) -> GlobalCounter {
+        match self {
+            Self::Read(global_counter, _) | Self::Write(global_counter, _) => *global_counter,
+        }
+    }
+
+    fn value(&self) -> Value<F> {
+        match self {
+            Self::Read(_, value) | Self::Write(_, value) => *value,
+        }
+    }
+
+    fn flag(&self) -> bool {
+        match self {
+            Self::Read(..) => false,
+            Self::Write(..) => true,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+/// All the read/write operations that happen at this address.
+pub(crate) struct Op<F: FieldExt> {
+    address: Address<F>,
+    global_counters: Vec<Option<ReadWrite<F>>>,
+}
+
+/// A mapping derived from witnessed memory operations.
+/// TODO: The complete version of this mapping will involve storage, stack,
+/// and opcode details as well.
+#[derive(Clone, Debug)]
+pub(crate) struct BusMapping<F: FieldExt> {
+    global_counter: Variable<usize, F>,
+    memory_flag: Variable<bool, F>,
+    memory_address: Variable<F, F>,
+    memory_value: Variable<F, F>,
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct Config<
+    F: FieldExt,
+    const GLOBAL_COUNTER_MAX: usize,
+    const MEMORY_ROWS_MAX: usize,
+    const MEMORY_ADDRESS_MAX: usize,
+    const STACK_ROWS_MAX: usize,
+    const STACK_ADDRESS_MAX: usize,
+> {
+    q_target: Column<Fixed>,
```

Comment:
Since we are using `q_target` to represent first row/stack/memory/storage, should we get rid of `q_first` and `q_not_first`?

Reply:
Yes, that's the idea. I already tried to remove these columns two but was having some problems. I will get back to this after I add the remaining constraints. However, using only q_target would probably require some more lookups because when you use q_not_first you know it's 0 or 1 and you don't need to normalize the value to 1 in range checks. On the other hand, q_target could be different values and we need to be careful (we need to normalize it) when checking whether this value multiplied by some other value is in some range. I plan to dig a bit into what's more efficient.

Reply:
@gaswhat q_first and and q_not_first have been removed and the number of lookups has been reduced. Hope it helps.

Reply:
I think `q_target` should be an advice column, since we'll have different numbers of different targets in each proof instance.

Reply:
There can be different numbers of different targets - there is a max number of memory, max number of stack, and max number of storage operations, but any number of operations up to the max value can be used. I think one of the main reasons I used a fixed column is because this way we can easily avoid the first row problem (which doesn't have Rot:prev). Instead, we could have one fixed column which would set a special value only for the first row, and q_target as an additional advice column - do you think this makes sense? It would complicate the constraints a bit though - for example, the first row could then be either memory, stack, or storage.